// zmp-react/cjs/components/routable-modals.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _useIsomorphicLayoutEffect = require("../shared/use-isomorphic-layout-effect");

var _zmp = require("../shared/zmp");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var RoutableModals = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var _useState = (0, _react.useState)([]),
      modals = _useState[0],
      _setModals = _useState[1];

  var elRef = (0, _react.useRef)(null);
  var routerData = (0, _react.useRef)(null);
  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current
    };
  });

  var onMount = function onMount() {
    routerData.current = {
      modals: modals,
      el: elRef.current,
      setModals: function setModals(newModals) {
        _setModals([].concat(newModals));
      }
    };
    _zmp.zmprouters.modals = routerData.current;
  };

  var onDestroy = function onDestroy() {
    if (!routerData.current) return;
    _zmp.zmprouters.modals = null;
    routerData.current = null;
  };

  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    onMount();
    return onDestroy;
  }, []);
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    if (!routerData.current || !_zmp.zmp) return;

    _zmp.zmpevents.emit('modalsRouterDidUpdate', routerData.current);
  });
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: elRef,
    className: "zmp-modals"
  }, modals.map(function (_ref) {
    var ModalComponent = _ref.component,
        modalId = _ref.id,
        modalProps = _ref.props;
    return /*#__PURE__*/_react.default.createElement(ModalComponent, (0, _extends2.default)({
      key: modalId
    }, modalProps));
  }));
});
RoutableModals.displayName = 'zmp-routable-modals';
var _default = RoutableModals;
exports.default = _default;
