// zmp-react/cjs/components/text-editor.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _useIsomorphicLayoutEffect = require("../shared/use-isomorphic-layout-effect");

var _utils = require("../shared/utils");

var _mixins = require("../shared/mixins");

var _zmp = require("../shared/zmp");

var _watchProp = require("../shared/watch-prop");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var TextEditor = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var _zmpTextEditor = (0, _react.useRef)(null);

  var className = props.className,
      id = props.id,
      style = props.style,
      mode = props.mode,
      value = props.value,
      buttons = props.buttons,
      customButtons = props.customButtons,
      dividers = props.dividers,
      imageUrlText = props.imageUrlText,
      linkUrlText = props.linkUrlText,
      placeholder = props.placeholder,
      clearFormattingOnPaste = props.clearFormattingOnPaste,
      _props$resizable = props.resizable,
      resizable = _props$resizable === void 0 ? false : _props$resizable;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);

  var onChange = function onChange(editor, editorValue) {
    (0, _utils.emit)(props, 'textEditorChange', editorValue);
  };

  var onInput = function onInput(editor, editorValue) {
    (0, _utils.emit)(props, 'textEditorInput', editorValue);
  };

  var onFocus = function onFocus() {
    (0, _utils.emit)(props, 'textEditorFocus');
  };

  var onBlur = function onBlur() {
    (0, _utils.emit)(props, 'textEditorBlur');
  };

  var onButtonClick = function onButtonClick(editor, button) {
    (0, _utils.emit)(props, 'textEditorButtonClick', button);
  };

  var onKeyboardOpen = function onKeyboardOpen() {
    (0, _utils.emit)(props, 'textEditorKeyboardOpen');
  };

  var onKeyboardClose = function onKeyboardClose() {
    (0, _utils.emit)(props, 'textEditorKeyboardClose');
  };

  var onPopoverOpen = function onPopoverOpen() {
    (0, _utils.emit)(props, 'textEditorPopoverOpen');
  };

  var onPopoverClose = function onPopoverClose() {
    (0, _utils.emit)(props, 'textEditorPopoverClose');
  };

  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current,
      zmpTextEditor: function zmpTextEditor() {
        return _zmpTextEditor.current;
      }
    };
  });
  (0, _watchProp.watchProp)(value, function (newValue) {
    if (_zmpTextEditor.current) {
      _zmpTextEditor.current.setValue(newValue);
    }
  });

  var onMount = function onMount() {
    var params = (0, _utils.noUndefinedProps)({
      el: elRef.current,
      mode: mode,
      value: value,
      buttons: buttons,
      customButtons: customButtons,
      dividers: dividers,
      imageUrlText: imageUrlText,
      linkUrlText: linkUrlText,
      placeholder: placeholder,
      clearFormattingOnPaste: clearFormattingOnPaste,
      on: {
        change: onChange,
        input: onInput,
        focus: onFocus,
        blur: onBlur,
        buttonClick: onButtonClick,
        keyboardOpen: onKeyboardOpen,
        keyboardClose: onKeyboardClose,
        popoverOpen: onPopoverOpen,
        popoverClose: onPopoverClose
      }
    });
    (0, _zmp.zmpready)(function () {
      _zmpTextEditor.current = _zmp.zmp.textEditor.create(params);
    });
  };

  var onDestroy = function onDestroy() {
    if (_zmpTextEditor.current && _zmpTextEditor.current.destroy) {
      _zmpTextEditor.current.destroy();
    }

    _zmpTextEditor.current = null;
  };

  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    onMount();
    return onDestroy;
  }, []);
  var slots = (0, _utils.getSlots)(props);
  var classes = (0, _utils.classNames)(className, 'text-editor', resizable && 'text-editor-resizable', (0, _mixins.colorClasses)(props));
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: elRef,
    id: id,
    style: style,
    className: classes
  }, extraAttrs), slots['root-start'], /*#__PURE__*/_react.default.createElement("div", {
    className: "text-editor-content",
    contentEditable: true
  }, slots.default), slots['root-end'], slots.root);
});
TextEditor.displayName = 'zmp-text-editor';
var _default = TextEditor;
exports.default = _default;
