// zmp-react/cjs/components/toolbar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _useIsomorphicLayoutEffect = require("../shared/use-isomorphic-layout-effect");

var _utils = require("../shared/utils");

var _mixins = require("../shared/mixins");

var _useTheme = require("../shared/use-theme");

var _zmp = require("../shared/zmp");

var _tabbarContext = require("../shared/tabbar-context");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var Toolbar = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var className = props.className,
      id = props.id,
      style = props.style,
      tabbar = props.tabbar,
      labels = props.labels,
      scrollable = props.scrollable,
      hidden = props.hidden,
      noShadow = props.noShadow,
      noHairline = props.noHairline,
      noBorder = props.noBorder,
      position = props.position,
      topMd = props.topMd,
      topIos = props.topIos,
      topAurora = props.topAurora,
      top = props.top,
      bottomMd = props.bottomMd,
      bottomIos = props.bottomIos,
      bottomAurora = props.bottomAurora,
      bottom = props.bottom,
      _props$inner = props.inner,
      inner = _props$inner === void 0 ? true : _props$inner;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);

  var onHide = function onHide(toolbarEl) {
    if (elRef.current !== toolbarEl) return;
    (0, _utils.emit)(props, 'toolbarHide');
  };

  var onShow = function onShow(toolbarEl) {
    if (elRef.current !== toolbarEl) return;
    (0, _utils.emit)(props, 'toolbarShow');
  };

  var hide = function hide(animate) {
    if (!_zmp.zmp) return;

    _zmp.zmp.toolbar.hide(elRef.current, animate);
  };

  var show = function show(animate) {
    if (!_zmp.zmp) return;

    _zmp.zmp.toolbar.show(elRef.current, animate);
  };

  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current,
      hide: hide,
      show: show
    };
  });
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    (0, _zmp.zmpready)(function () {
      if (tabbar && _zmp.zmp && elRef.current) {
        _zmp.zmp.toolbar.setHighlight(elRef.current);
      }

      _zmp.zmp.on('toolbarShow', onShow);

      _zmp.zmp.on('toolbarHide', onHide);
    });
    return function () {
      if (!_zmp.zmp) return;

      _zmp.zmp.off('toolbarShow', onShow);

      _zmp.zmp.off('toolbarHide', onHide);
    };
  });
  var theme = (0, _useTheme.useTheme)();
  var classes = (0, _utils.classNames)(className, 'toolbar', {
    tabbar: tabbar,
    'toolbar-bottom': theme && theme.md && bottomMd || theme && theme.ios && bottomIos || theme && theme.aurora && bottomAurora || bottom || position === 'bottom',
    'toolbar-top': theme && theme.md && topMd || theme && theme.ios && topIos || theme && theme.aurora && topAurora || top || position === 'top',
    'tabbar-labels': labels,
    'tabbar-scrollable': scrollable,
    'toolbar-hidden': hidden,
    'no-shadow': noShadow,
    'no-hairline': noHairline || noBorder
  }, (0, _mixins.colorClasses)(props));
  var slots = (0, _utils.getSlots)(props);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), /*#__PURE__*/_react.default.createElement(_tabbarContext.TabbarContext.Provider, {
    value: {
      tabbarHasLabels: labels
    }
  }, slots['before-inner'], inner ? /*#__PURE__*/_react.default.createElement("div", {
    className: "toolbar-inner"
  }, slots.default) : slots.default, slots['after-inner']));
});
Toolbar.displayName = 'zmp-toolbar';
var _default = Toolbar;
exports.default = _default;
