// zmp-react/esm/zmp-react.js
/**
 * ZMP React 2.8.10
 * Build full featured iOS & Android apps using ZMP & React
 * https://h5.zalo.me/react/
 *
 * Copyright 2014-2023 Vladimir Kharlampidi
 *
 * Released under the MIT License
 *
 * Released on: May 16, 2023
 */
/* eslint-disable no-proto */
/* eslint-disable func-names */
import ZMPReact from './shared/plugin';
import { zmp, zmpready, theme } from './shared/zmp';
import { useStore } from './shared/use-store';

import ActionsButton from './zalo-ui-kits/actions-button';
import ActionsGroup from './zalo-ui-kits/actions-group';
import ActionsLabel from './zalo-ui-kits/actions-label';
import Actions from './zalo-ui-kits/actions';
import App from './zalo-ui-kits/app';
import AvatarGroup from './zalo-ui-kits/avatar-group';
import Avatar from './zalo-ui-kits/avatar';
import Box from './zalo-ui-kits/box';
import Button from './zalo-ui-kits/button';
import Card from './zalo-ui-kits/card';
import Checkbox from './zalo-ui-kits/checkbox';
import Col from './zalo-ui-kits/col';
import DatePicker from './zalo-ui-kits/date-picker';
import FabBackdrop from './zalo-ui-kits/fab-backdrop';
import FabButton from './zalo-ui-kits/fab-button';
import FabButtons from './zalo-ui-kits/fab-buttons';
import Fab from './zalo-ui-kits/fab';
import GridItem from './zalo-ui-kits/grid-item';
import Grid from './zalo-ui-kits/grid';
import Icon from './zalo-ui-kits/icon';
import InputInfo from './zalo-ui-kits/input-info';
import Input from './zalo-ui-kits/input';
import Link from './zalo-ui-kits/link';
import ListInput from './zalo-ui-kits/list-input';
import ListItem from './zalo-ui-kits/list-item';
import List from './zalo-ui-kits/list';
import NavLeft from './zalo-ui-kits/nav-left';
import NavRight from './zalo-ui-kits/nav-right';
import NavTitleLarge from './zalo-ui-kits/nav-title-large';
import NavTitle from './zalo-ui-kits/nav-title';
import Navbar from './zalo-ui-kits/navbar';
import Page from './zalo-ui-kits/page';
import Picker from './zalo-ui-kits/picker';
import Preloader from './zalo-ui-kits/preloader';
import Progressbar from './zalo-ui-kits/progressbar';
import Radio from './zalo-ui-kits/radio';
import Range from './zalo-ui-kits/range';
import Row from './zalo-ui-kits/row';
import Searchbar from './zalo-ui-kits/searchbar';
import Sheet from './zalo-ui-kits/sheet';
import SkeletonAvatar from './zalo-ui-kits/skeleton-avatar';
import SkeletonBlock from './zalo-ui-kits/skeleton-block';
import SkeletonImage from './zalo-ui-kits/skeleton-image';
import SkeletonText from './zalo-ui-kits/skeleton-text';
import SwiperSlide from './zalo-ui-kits/swiper-slide';
import Swiper from './zalo-ui-kits/swiper';
import TabView from './zalo-ui-kits/tab-view';
import Tab from './zalo-ui-kits/tab';
import Tabbar from './zalo-ui-kits/tabbar';
import Tabs from './zalo-ui-kits/tabs';
import Text from './zalo-ui-kits/text';
import Title from './zalo-ui-kits/title';
import ToastPreloader from './zalo-ui-kits/toast-preloader';
import Toggle from './zalo-ui-kits/toggle';
import View from './zalo-ui-kits/view';

import windowPlugin from './shared/window';

windowPlugin();

export { ActionsButton, ActionsGroup, ActionsLabel, Actions, App, AvatarGroup, Avatar, Box, Button, Card, Checkbox, Col, DatePicker, FabBackdrop, FabButton, FabButtons, Fab, GridItem, Grid, Icon, InputInfo, Input, Link, ListInput, ListItem, List, NavLeft, NavRight, NavTitleLarge, NavTitle, Navbar, Page, Picker, Preloader, Progressbar, Radio, Range, Row, Searchbar, Sheet, SkeletonAvatar, SkeletonBlock, SkeletonImage, SkeletonText, SwiperSlide, Swiper, TabView, Tab, Tabbar, Tabs, Text, Title, ToastPreloader, Toggle, View }
export { zmp, zmpready, theme, useStore };
export default ZMPReact;
