// zmp-react/esm/zalo-ui-kits/list-input.js
import _extends from "@babel/runtime/helpers/extends";
import React, { useState, forwardRef, useRef, useImperativeHandle, useMemo } from 'react';
import ListInput from '../components/list-input';
import InputInfo from './input-info';
import Icon from './icon';
import { ICON_SIZE_MD, ICON_EYE, ICON_EYE_FILL, RESIZABLE_TEXTAREA_HEIGHT } from '../../common/constants';
import { classNames } from '../shared/utils';
import { isNumber } from '../../common/utils';
/* dts-imports
import { Calendar, ColorPicker, TextEditor } from 'zmp-core/types';
*/

/* dts-props
  id?: string | number;
  className?: string;
  style?: React.CSSProperties;
  wrap? : boolean
  input? : boolean
  type? : string
  name? : string
  value? : string | number | Array<any> | Date | Object
  defaultValue? : string | number | Array<any>
  inputmode? : string
  readonly? : boolean
  required? : boolean
  disabled? : boolean
  placeholder? : string
  inputId? : string | number
  size? : string | number
  accept? : string | number
  autocomplete? : string
  autocorrect? : string
  autocapitalize? : string
  spellcheck? : string
  autofocus? : boolean
  autosave? : string
  max? : string | number
  min? : string | number
  step? : string | number
  maxlength? : string | number
  minlength? : string | number
  multiple? : boolean
  inputStyle?: React.CSSProperties;
  pattern? : string
  validate? : boolean | string
  validateOnBlur? : boolean
  onValidate? : Function
  tabindex? : string | number
  resizable? : boolean
  minRows? : number
  maxRows: number
  clearButton? : boolean
  noFormStoreData? : boolean
  noStoreData? : boolean
  ignoreStoreData? : boolean
  errorMessage? : string
  errorMessageForce? : boolean
  info? : string
  label? : string | number
  calendarParams? : Calendar.Parameters;
  colorPickerParams? : ColorPicker.Parameters;
  textEditorParams? : TextEditor.Parameters;
  COLOR_PROPS
  onCalendarChange? : (calendarValue?: any) => void
  onColorPickerChange? : (colorPickerValue?: any) => void
  onTextareaResize? : (event?: any) => void
  onInputNotEmpty? : (event?: any) => void
  onInputEmpty? : (event?: any) => void
  onInputClear? : (event?: any) => void
  onInput? : (...args: any[]) => void
  onFocus? : (...args: any[]) => void
  onBlur? : (...args: any[]) => void
  onChange? : (...args: any[]) => void
  onTextEditorChange? : (...args: any[]) => void
  ref?: React.MutableRefObject<{el: HTMLElement | null}>;
  CHILDREN_PROP
*/

var ZMPInput = /*#__PURE__*/forwardRef(function (props, ref) {
  var _classNames;

  var elRef = useRef(null);

  var _useState = useState(undefined),
      valid = _useState[0],
      setValid = _useState[1];

  var _useState2 = useState(false),
      showPassword = _useState2[0],
      setShowPassword = _useState2[1];

  var _onValidate = props.onValidate,
      info = props.info,
      type = props.type,
      minRows = props.minRows,
      _props$maxRows = props.maxRows,
      maxRows = _props$maxRows === void 0 ? 8 : _props$maxRows;
  useImperativeHandle(ref, function () {
    return elRef.current;
  });

  var handleOnValidate = function handleOnValidate(isValid) {
    setValid(isValid);
  };

  var handleClickShowPassword = function handleClickShowPassword() {
    setShowPassword(!showPassword);
  };

  var isValidTextareaMinRows = useMemo(function () {
    return isNumber(minRows) && minRows >= RESIZABLE_TEXTAREA_HEIGHT.MIN_ROWS.MIN && minRows <= RESIZABLE_TEXTAREA_HEIGHT.MIN_ROWS.MAX;
  }, [minRows]);
  var isValidTextareaMaxRows = useMemo(function () {
    return isNumber(maxRows) && maxRows >= RESIZABLE_TEXTAREA_HEIGHT.MAX_ROWS.MIN && maxRows <= RESIZABLE_TEXTAREA_HEIGHT.MAX_ROWS.MAX;
  }, [maxRows]);
  var inputClasses = classNames((_classNames = {
    'item-input-successful': valid
  }, _classNames["textarea-rows-" + minRows] = type === 'textarea' && props.resizable && isValidTextareaMinRows, _classNames["textarea-max-rows-" + maxRows] = type === 'textarea' && props.resizable && isValidTextareaMaxRows, _classNames));

  if (type === 'password') {
    return /*#__PURE__*/React.createElement(ListInput, _extends({
      ref: elRef
    }, props, {
      sortable: false,
      media: null,
      inlineLabel: false,
      floatingLabel: false,
      outline: false,
      className: "item-input-password " + (valid ? 'item-input-successful' : ''),
      info: /*#__PURE__*/React.createElement(InputInfo, {
        info: info,
        valid: valid
      }),
      type: showPassword ? 'text' : 'password',
      onValidate: function onValidate(isValid) {
        handleOnValidate(isValid);

        if (_onValidate && typeof _onValidate === 'function') {
          _onValidate(isValid);
        }
      }
    }), /*#__PURE__*/React.createElement("span", {
      slot: "input",
      onClick: handleClickShowPassword,
      className: "show-password-button"
    }, /*#__PURE__*/React.createElement(Icon, {
      zmp: "" + (showPassword ? ICON_EYE_FILL : ICON_EYE),
      size: ICON_SIZE_MD
    })));
  }

  return /*#__PURE__*/React.createElement(ListInput, _extends({}, props, {
    sortable: false,
    media: null,
    inlineLabel: false,
    outline: false,
    className: inputClasses,
    rows: type === 'textarea' && isValidTextareaMinRows ? minRows : undefined,
    info: /*#__PURE__*/React.createElement(InputInfo, {
      info: info,
      valid: valid
    }),
    onValidate: function onValidate(isValid) {
      handleOnValidate(isValid);

      if (_onValidate && typeof _onValidate === 'function') {
        _onValidate(isValid);
      }
    }
  }));
});
ZMPInput.displayName = 'zmp-list-input';
export default ZMPInput;
