// zmp-react/esm/components/actions.js
import _extends from "@babel/runtime/helpers/extends";
import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { useIsomorphicLayoutEffect } from '../shared/use-isomorphic-layout-effect';
import { classNames, getExtraAttrs, emit } from '../shared/utils';
import { colorClasses } from '../shared/mixins';
import { zmp, zmpready } from '../shared/zmp';
import { watchProp } from '../shared/watch-prop';
import { modalStateClasses } from '../shared/modal-state-classes';


var Actions = /*#__PURE__*/forwardRef(function (props, ref) {
  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      grid = props.grid,
      _props$opened = props.opened,
      opened = _props$opened === void 0 ? false : _props$opened,
      animate = props.animate;
  var extraAttrs = getExtraAttrs(props);
  var elRef = useRef(null);
  var isOpened = useRef(opened);
  var isClosing = useRef(false);

  var _zmpActions = useRef(null);

  var onOpen = function onOpen(instance) {
    isOpened.current = true;
    isClosing.current = false;
    emit(props, 'actionsOpen', instance);
  };

  var onOpened = function onOpened(instance) {
    emit(props, 'actionsOpened', instance);
  };

  var onClose = function onClose(instance) {
    isOpened.current = false;
    isClosing.current = true;
    emit(props, 'actionsClose', instance);
  };

  var onClosed = function onClosed(instance) {
    isClosing.current = false;
    emit(props, 'actionsClosed', instance);
  };

  useImperativeHandle(ref, function () {
    return {
      el: elRef.current,
      zmpActions: function zmpActions() {
        return _zmpActions.current;
      }
    };
  }); // watch opened changes

  watchProp(opened, function (value) {
    if (!_zmpActions.current) return;

    if (value) {
      _zmpActions.current.open();
    } else {
      _zmpActions.current.close();
    }
  });

  var onMount = function onMount() {
    if (!elRef.current) return;
    var target = props.target,
        convertToPopover = props.convertToPopover,
        forceToPopover = props.forceToPopover,
        closeByBackdropClick = props.closeByBackdropClick,
        closeByOutsideClick = props.closeByOutsideClick,
        closeOnEscape = props.closeOnEscape,
        backdrop = props.backdrop,
        backdropEl = props.backdropEl,
        containerEl = props.containerEl;
    var params = {
      el: elRef.current,
      grid: grid,
      on: {
        open: onOpen,
        opened: onOpened,
        close: onClose,
        closed: onClosed
      }
    };
    if (target) params.targetEl = target;
    if ('convertToPopover' in props) params.convertToPopover = convertToPopover;
    if ('forceToPopover' in props) params.forceToPopover = forceToPopover;
    if ('backdrop' in props) params.backdrop = backdrop;
    if ('backdropEl' in props) params.backdropEl = backdropEl;
    if ('closeByBackdropClick' in props) params.closeByBackdropClick = closeByBackdropClick;
    if ('closeByOutsideClick' in props) params.closeByOutsideClick = closeByOutsideClick;
    if ('closeOnEscape' in props) params.closeOnEscape = closeOnEscape;
    if ('animate' in props) params.animate = animate;
    if ('containerEl' in props) params.containerEl = containerEl;
    zmpready(function () {
      _zmpActions.current = zmp.actions.create(params);

      if (opened) {
        _zmpActions.current.open(false);
      }
    });
  };

  var onDestroy = function onDestroy() {
    if (_zmpActions.current) {
      if (_zmpActions.current.opened) {
        _zmpActions.current.close();
      } else {
        _zmpActions.current.destroy();
      }
    }

    _zmpActions.current = null;
  };

  useIsomorphicLayoutEffect(function () {
    onMount();
    return onDestroy;
  }, []);
  var classes = classNames(className, 'actions-modal', {
    'actions-grid': grid
  }, modalStateClasses({
    isOpened: isOpened,
    isClosing: isClosing
  }), colorClasses(props));
  return /*#__PURE__*/React.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), children);
});
Actions.displayName = 'zmp-actions';
export default Actions;
