// zmp-react/esm/components/messages.js
import _extends from "@babel/runtime/helpers/extends";
import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { useIsomorphicLayoutEffect } from '../shared/use-isomorphic-layout-effect';
import { classNames, getExtraAttrs, noUndefinedProps } from '../shared/utils';
import { colorClasses } from '../shared/mixins';
import { zmpready, zmp } from '../shared/zmp';
import { watchProp } from '../shared/watch-prop';

var Messages = /*#__PURE__*/forwardRef(function (props, ref) {
  var _zmpMessages = useRef(null);

  var mounted = useRef(false);
  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      _props$autoLayout = props.autoLayout,
      autoLayout = _props$autoLayout === void 0 ? false : _props$autoLayout,
      _props$messages = props.messages,
      messages = _props$messages === void 0 ? [] : _props$messages,
      _props$newMessagesFir = props.newMessagesFirst,
      newMessagesFirst = _props$newMessagesFir === void 0 ? false : _props$newMessagesFir,
      _props$scrollMessages = props.scrollMessages,
      scrollMessages = _props$scrollMessages === void 0 ? true : _props$scrollMessages,
      _props$scrollMessages2 = props.scrollMessagesOnEdge,
      scrollMessagesOnEdge = _props$scrollMessages2 === void 0 ? true : _props$scrollMessages2,
      firstMessageRule = props.firstMessageRule,
      lastMessageRule = props.lastMessageRule,
      tailMessageRule = props.tailMessageRule,
      sameNameMessageRule = props.sameNameMessageRule,
      sameHeaderMessageRule = props.sameHeaderMessageRule,
      sameFooterMessageRule = props.sameFooterMessageRule,
      sameAvatarMessageRule = props.sameAvatarMessageRule,
      customClassMessageRule = props.customClassMessageRule,
      renderMessage = props.renderMessage,
      _props$typing = props.typing,
      typing = _props$typing === void 0 ? false : _props$typing,
      _props$init = props.init,
      init = _props$init === void 0 ? true : _props$init;
  var extraAttrs = getExtraAttrs(props);
  var elRef = useRef(null);
  var childrenBeforeUpdated = useRef(null);
  useImperativeHandle(ref, function () {
    return {
      el: elRef.current,
      zmpMessages: function zmpMessages() {
        return _zmpMessages.current;
      }
    };
  });

  var onMount = function onMount() {
    if (!init) return;
    zmpready(function () {
      _zmpMessages.current = zmp.messages.create(noUndefinedProps({
        el: elRef.current,
        autoLayout: autoLayout,
        messages: messages,
        newMessagesFirst: newMessagesFirst,
        scrollMessages: scrollMessages,
        scrollMessagesOnEdge: scrollMessagesOnEdge,
        firstMessageRule: firstMessageRule,
        lastMessageRule: lastMessageRule,
        tailMessageRule: tailMessageRule,
        sameNameMessageRule: sameNameMessageRule,
        sameHeaderMessageRule: sameHeaderMessageRule,
        sameFooterMessageRule: sameFooterMessageRule,
        sameAvatarMessageRule: sameAvatarMessageRule,
        customClassMessageRule: customClassMessageRule,
        renderMessage: renderMessage
      }));

      if (typing) {
        _zmpMessages.current.showTyping();
      }
    });
  };

  var onDestroy = function onDestroy() {
    if (_zmpMessages.current && _zmpMessages.current.destroy) _zmpMessages.current.destroy();
    _zmpMessages.current = null;
  };

  useIsomorphicLayoutEffect(function () {
    onMount();
    return onDestroy;
  }, []);
  useIsomorphicLayoutEffect(function () {
    var wasMounted = mounted.current;
    mounted.current = true;
    if (!init || !elRef.current) return;
    var childElements = elRef.current.children;
    if (!childElements) return;
    var childrenAfterUpdated = childElements.length;

    if (!wasMounted) {
      for (var i = 0; i < childElements.length; i += 1) {
        childElements[i].classList.add('message-appeared');
      }

      return;
    }

    for (var _i = 0; _i < childElements.length; _i += 1) {
      if (!childElements[_i].classList.contains('message-appeared')) {
        childElements[_i].classList.add('message-appear-from-bottom');
      }
    }

    if (_zmpMessages.current) {
      if (_zmpMessages.current.layout && autoLayout) {
        _zmpMessages.current.layout();
      }

      if (childrenBeforeUpdated.current !== childrenAfterUpdated && _zmpMessages.current.scroll && scrollMessages) {
        _zmpMessages.current.scroll();
      }
    }

    childrenBeforeUpdated.current = childrenAfterUpdated;
  });
  watchProp(typing, function (newValue) {
    if (!_zmpMessages.current) return;
    if (newValue) _zmpMessages.current.showTyping();else _zmpMessages.current.hideTyping();
  });
  var classes = classNames(className, 'messages', colorClasses(props));
  return /*#__PURE__*/React.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), children);
});
Messages.displayName = 'zmp-messages';
export default Messages;
