// zmp-react/esm/components/nav-title.js
import _extends from "@babel/runtime/helpers/extends";
import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { classNames, getExtraAttrs } from '../shared/utils';
import { colorClasses } from '../shared/mixins';

var NavTitle = /*#__PURE__*/forwardRef(function (props, ref) {
  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      title = props.title,
      subtitle = props.subtitle,
      sliding = props.sliding;
  var extraAttrs = getExtraAttrs(props);
  var elRef = useRef(null);
  useImperativeHandle(ref, function () {
    return {
      el: elRef.current
    };
  });
  var subtitleEl;

  if (subtitle) {
    subtitleEl = /*#__PURE__*/React.createElement("span", {
      className: "subtitle"
    }, subtitle);
  }

  var classes = classNames(className, 'title', {
    sliding: sliding
  }, colorClasses(props));
  return /*#__PURE__*/React.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), children, title, subtitleEl);
});
NavTitle.displayName = 'zmp-nav-title';
export default NavTitle;
