// zmp-react/esm/components/panel.js
import _extends from "@babel/runtime/helpers/extends";
import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { useIsomorphicLayoutEffect } from '../shared/use-isomorphic-layout-effect';
import { classNames, getExtraAttrs, noUndefinedProps, emit } from '../shared/utils';
import { colorClasses } from '../shared/mixins';
import { zmpready, zmp } from '../shared/zmp';
import { watchProp } from '../shared/watch-prop';

var Panel = /*#__PURE__*/forwardRef(function (props, ref) {
  var _classNames;

  var _zmpPanel = useRef(null);

  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      side = props.side,
      effect = props.effect,
      reveal = props.reveal,
      left = props.left,
      opened = props.opened,
      resizable = props.resizable,
      _props$backdrop = props.backdrop,
      backdrop = _props$backdrop === void 0 ? true : _props$backdrop,
      backdropEl = props.backdropEl,
      containerEl = props.containerEl,
      visibleBreakpoint = props.visibleBreakpoint,
      collapsedBreakpoint = props.collapsedBreakpoint,
      swipe = props.swipe,
      swipeNoFollow = props.swipeNoFollow,
      swipeOnlyClose = props.swipeOnlyClose,
      _props$swipeActiveAre = props.swipeActiveArea,
      swipeActiveArea = _props$swipeActiveAre === void 0 ? 0 : _props$swipeActiveAre,
      _props$swipeThreshold = props.swipeThreshold,
      swipeThreshold = _props$swipeThreshold === void 0 ? 0 : _props$swipeThreshold;
  var extraAttrs = getExtraAttrs(props);
  var elRef = useRef(null);
  var isOpened = useRef(false);
  var isClosing = useRef(false);
  var isCollapsed = useRef(false);
  var isBreakpoint = useRef(false);

  var onOpen = function onOpen(event) {
    isOpened.current = true;
    isClosing.current = false;
    emit(props, 'panelOpen', event);
  };

  var onOpened = function onOpened(event) {
    emit(props, 'panelOpened', event);
  };

  var onClose = function onClose(event) {
    isOpened.current = false;
    isClosing.current = true;
    emit(props, 'panelClose', event);
  };

  var onClosed = function onClosed(event) {
    isClosing.current = false;
    emit(props, 'panelClosed', event);
  };

  var onBackdropClick = function onBackdropClick(event) {
    emit(props, 'click panelBackdropClick', event);
  };

  var onSwipe = function onSwipe(event) {
    emit(props, 'panelSwipe', event);
  };

  var onSwipeOpen = function onSwipeOpen(event) {
    emit(props, 'panelSwipeOpen', event);
  };

  var onBreakpoint = function onBreakpoint(event) {
    isBreakpoint.current = true;
    isCollapsed.current = false;
    emit(props, 'panelBreakpoint', event);
  };

  var onCollapsedBreakpoint = function onCollapsedBreakpoint(event) {
    isBreakpoint.current = false;
    isCollapsed.current = true;
    emit(props, 'panelCollapsedBreakpoint', event);
  };

  var onResize = function onResize() {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    emit.apply(void 0, [props, 'panelResize'].concat(args));
  };

  useImperativeHandle(ref, function () {
    return {
      el: elRef.current,
      zmpPanel: function zmpPanel() {
        return _zmpPanel.current;
      }
    };
  });
  watchProp(resizable, function (newValue) {
    if (!_zmpPanel.current) return;
    if (newValue) _zmpPanel.current.enableResizable();else _zmpPanel.current.disableResizable();
  });
  watchProp(opened, function (newValue) {
    if (!_zmpPanel.current) return;

    if (newValue) {
      _zmpPanel.current.open();
    } else {
      _zmpPanel.current.close();
    }
  });

  var onMount = function onMount() {
    zmpready(function () {
      var $ = zmp.$;
      if (!$) return;

      if ($('.panel-backdrop').length === 0) {
        $('<div class="panel-backdrop"></div>').insertBefore(elRef.current);
      }

      var params = noUndefinedProps({
        el: elRef.current,
        resizable: resizable,
        backdrop: backdrop,
        backdropEl: backdropEl,
        containerEl: containerEl,
        visibleBreakpoint: visibleBreakpoint,
        collapsedBreakpoint: collapsedBreakpoint,
        swipe: swipe,
        swipeNoFollow: swipeNoFollow,
        swipeOnlyClose: swipeOnlyClose,
        swipeActiveArea: swipeActiveArea,
        swipeThreshold: swipeThreshold,
        on: {
          open: onOpen,
          opened: onOpened,
          close: onClose,
          closed: onClosed,
          backdropClick: onBackdropClick,
          swipe: onSwipe,
          swipeOpen: onSwipeOpen,
          collapsedBreakpoint: onCollapsedBreakpoint,
          breakpoint: onBreakpoint,
          resize: onResize
        }
      });
      _zmpPanel.current = zmp.panel.create(params);

      if (opened) {
        _zmpPanel.current.open(false);
      }
    });
  };

  var onDestroy = function onDestroy() {
    if (_zmpPanel.current && _zmpPanel.current.destroy) {
      _zmpPanel.current.destroy();
    }

    _zmpPanel.current = null;
  };

  useIsomorphicLayoutEffect(function () {
    onMount();
    return onDestroy;
  }, []);
  var sideComputed = side || (left ? 'left' : 'right');
  var effectComputed = effect || (reveal ? 'reveal' : 'cover');
  var classes = classNames(className, 'panel', (_classNames = {
    'panel-in': isOpened.current && !isClosing.current && !isBreakpoint.current,
    'panel-in-breakpoint': isBreakpoint.current,
    'panel-in-collapsed': isCollapsed.current,
    'panel-resizable': resizable
  }, _classNames["panel-" + sideComputed] = sideComputed, _classNames["panel-" + effectComputed] = effectComputed, _classNames), colorClasses(props));
  return /*#__PURE__*/React.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), children, resizable && /*#__PURE__*/React.createElement("div", {
    className: "panel-resize-handler"
  }));
});
Panel.displayName = 'zmp-panel';
export default Panel;
