// zmp-react/esm/components/subnavbar.js
import _extends from "@babel/runtime/helpers/extends";
import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { classNames, getExtraAttrs } from '../shared/utils';
import { colorClasses } from '../shared/mixins';

var Subnavbar = /*#__PURE__*/forwardRef(function (props, ref) {
  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      _props$inner = props.inner,
      inner = _props$inner === void 0 ? true : _props$inner,
      title = props.title,
      sliding = props.sliding;
  var extraAttrs = getExtraAttrs(props);
  var elRef = useRef(null);
  useImperativeHandle(ref, function () {
    return {
      el: elRef.current
    };
  });
  var classes = classNames(className, 'subnavbar', {
    sliding: sliding
  }, colorClasses(props));
  return /*#__PURE__*/React.createElement("div", _extends({
    className: classes,
    id: id,
    style: style,
    ref: elRef
  }, extraAttrs), inner ? /*#__PURE__*/React.createElement("div", {
    className: "subnavbar-inner"
  }, title && /*#__PURE__*/React.createElement("div", {
    className: "subnavbar-title"
  }, title), children) : children);
});
Subnavbar.displayName = 'zmp-subnavbar';
export default Subnavbar;
