// zmp-react/esm/components/swipeout-actions.js
import _extends from "@babel/runtime/helpers/extends";
import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { classNames, getExtraAttrs } from '../shared/utils';
import { colorClasses } from '../shared/mixins';

var SwipeoutActions = /*#__PURE__*/forwardRef(function (props, ref) {
  var className = props.className,
      id = props.id,
      style = props.style,
      children = props.children,
      left = props.left,
      right = props.right,
      side = props.side;
  var extraAttrs = getExtraAttrs(props);
  var elRef = useRef(null);
  useImperativeHandle(ref, function () {
    return {
      el: elRef.current
    };
  });
  var sideComputed = side;

  if (!sideComputed) {
    if (left) sideComputed = 'left';
    if (right) sideComputed = 'right';
  }

  var classes = classNames(className, "swipeout-actions-" + sideComputed, colorClasses(props));
  return /*#__PURE__*/React.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), children);
});
SwipeoutActions.displayName = 'zmp-swipeout-actions';
export default SwipeoutActions;
