// zmp-core/cjs/components/view/view.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _zmpDom = _interopRequireDefault(require("../../shared/zmp-dom"));

var _utils = require("../../shared/utils");

var _viewClass = _interopRequireDefault(require("./view-class"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getCurrentView(app) {
  var $popoverView = (0, _zmpDom.default)('.popover.modal-in .view');
  var $popupView = (0, _zmpDom.default)('.popup.modal-in .view');
  var $panelView = (0, _zmpDom.default)('.panel.panel-in .view');
  var $viewsEl = (0, _zmpDom.default)('.views');
  if ($viewsEl.length === 0) $viewsEl = app.$el; // Find active view as tab

  var $viewEl = $viewsEl.children('.view');

  if ($viewEl.length === 0) {
    $viewEl = $viewsEl.children('.tabs').children('.view');
  } // Propably in tabs or split view


  if ($viewEl.length > 1) {
    if ($viewEl.hasClass('tab')) {
      // Tabs
      $viewEl = $viewsEl.children('.view.tab-active');

      if ($viewEl.length === 0) {
        $viewEl = $viewsEl.children('.tabs').children('.view.tab-active');
      }
    } else {// Split View, leave appView intact
    }
  }

  if ($popoverView.length > 0 && $popoverView[0].zmpView) return $popoverView[0].zmpView;
  if ($popupView.length > 0 && $popupView[0].zmpView) return $popupView[0].zmpView;
  if ($panelView.length > 0 && $panelView[0].zmpView) return $panelView[0].zmpView;

  if ($viewEl.length > 0) {
    if ($viewEl.length === 1 && $viewEl[0].zmpView) return $viewEl[0].zmpView;

    if ($viewEl.length > 1) {
      return app.views.main;
    }
  }

  return undefined;
}

var _default = {
  name: 'view',
  params: {
    view: {
      init: true,
      initRouterOnTabShow: false,
      name: undefined,
      main: false,
      router: true,
      linksView: null,
      stackPages: false,
      xhrCache: true,
      xhrCacheIgnore: [],
      xhrCacheIgnoreGetParameters: false,
      xhrCacheDuration: 1000 * 60 * 10,
      // Ten minutes
      componentCache: true,
      preloadPreviousPage: false,
      allowDuplicateUrls: false,
      reloadPages: false,
      reloadDetail: false,
      masterDetailBreakpoint: 0,
      masterDetailResizable: false,
      removeElements: true,
      removeElementsWithTimeout: false,
      removeElementsTimeout: 0,
      restoreScrollTopOnBack: true,
      unloadTabContent: true,
      passRouteQueryToRequest: true,
      passRouteParamsToRequest: false,
      loadInitialPage: true,
      // Swipe Back
      iosSwipeBack: true,
      iosSwipeBackAnimateShadow: true,
      iosSwipeBackAnimateOpacity: true,
      iosSwipeBackActiveArea: 30,
      iosSwipeBackThreshold: 0,
      mdSwipeBack: false,
      mdSwipeBackAnimateShadow: true,
      mdSwipeBackAnimateOpacity: false,
      mdSwipeBackActiveArea: 30,
      mdSwipeBackThreshold: 0,
      auroraSwipeBack: false,
      auroraSwipeBackAnimateShadow: false,
      auroraSwipeBackAnimateOpacity: true,
      auroraSwipeBackActiveArea: 30,
      auroraSwipeBackThreshold: 0,
      // Push State
      browserHistory: true,
      // eslint-disable-next-line no-restricted-globals
      browserHistoryRoot: process.env.NODE_ENV === 'production' ? "/zapps/" + window.APP_ID : '#!',
      browserHistoryAnimate: false,
      browserHistoryAnimateOnLoad: false,
      browserHistorySeparator: '',
      browserHistoryOnLoad: true,
      browserHistoryInitialMatch: false,
      browserHistoryStoreHistory: false,
      // Animate Pages
      animate: true,
      // iOS Dynamic Navbar
      iosDynamicNavbar: false,
      // Animate iOS Navbar Back Icon
      iosAnimateNavbarBackIcon: true,
      // Delays
      iosPageLoadDelay: 0,
      mdPageLoadDelay: 0,
      auroraPageLoadDelay: 0,
      // Routes hooks
      routesBeforeEnter: null,
      routesBeforeLeave: null
    }
  },
  static: {
    View: _viewClass.default
  },
  create: function create() {
    var app = this;
    (0, _utils.extend)(app, {
      views: (0, _utils.extend)([], {
        create: function create(el, params) {
          return new _viewClass.default(app, el, params);
        },
        get: function get(viewEl) {
          var $viewEl = (0, _zmpDom.default)(viewEl);
          if ($viewEl.length && $viewEl[0].zmpView) return $viewEl[0].zmpView;
          return undefined;
        }
      })
    });
    Object.defineProperty(app.views, 'current', {
      enumerable: true,
      configurable: true,
      get: function get() {
        return getCurrentView(app);
      }
    }); // Alias

    app.view = app.views;
  },
  on: {
    init: function init() {
      var app = this;
      (0, _zmpDom.default)('.view-init').each(function (viewEl) {
        if (viewEl.zmpView) return;
        var viewParams = (0, _zmpDom.default)(viewEl).dataset();
        app.views.create(viewEl, viewParams);
      });
    },
    'modalOpen panelOpen': function onOpen(instance) {
      var app = this;
      instance.$el.find('.view-init').each(function (viewEl) {
        if (viewEl.zmpView) return;
        var viewParams = (0, _zmpDom.default)(viewEl).dataset();
        app.views.create(viewEl, viewParams);
      });
    },
    'modalBeforeDestroy panelBeforeDestroy': function onClose(instance) {
      if (!instance || !instance.$el) return;
      instance.$el.find('.view-init').each(function (viewEl) {
        var view = viewEl.zmpView;
        if (!view) return;
        view.destroy();
      });
    }
  },
  vnode: {
    'view-init': {
      insert: function insert(vnode) {
        var app = this;
        var viewEl = vnode.elm;
        if (viewEl.zmpView) return;
        var viewParams = (0, _zmpDom.default)(viewEl).dataset();
        app.views.create(viewEl, viewParams);
      },
      destroy: function destroy(vnode) {
        var viewEl = vnode.elm;
        var view = viewEl.zmpView;
        if (!view) return;
        view.destroy();
      }
    }
  }
};
exports.default = _default;
