// zmp-react/cjs/zalo-ui-kits/avatar-group.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _utils = require("../shared/utils");

var _constants = require("../../common/constants");

var _avatar = _interopRequireDefault(require("./avatar"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* dts-props
  id: string | number;
  className: string;
  style: React.CSSProperties;
  maxCounter? : number;
  horizontal?: boolean;
  onCounterClick?: (event?: any) => void
  ref?: React.MutableRefObject<{el: HTMLElement | null}>;
  CHILDREN_PROP
*/
var AvatarGroup = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var id = _ref.id,
      children = _ref.children,
      horizontalProps = _ref.horizontal,
      className = _ref.className,
      style = _ref.style,
      _ref$maxCounter = _ref.maxCounter,
      maxCounterProps = _ref$maxCounter === void 0 ? _constants.DEFAULT_AVATAR_GROUP_COUNTER : _ref$maxCounter,
      onCounterClick = _ref.onCounterClick;
  var elRef = (0, _react.useRef)(null);
  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current
    };
  });

  var handleOnCounterClick = function handleOnCounterClick(e) {
    if (typeof onCounterClick === 'function') {
      onCounterClick(e);
    }
  };

  var maxCounter = maxCounterProps < _constants.DEFAULT_AVATAR_GROUP_COUNTER ? _constants.DEFAULT_AVATAR_GROUP_COUNTER : maxCounterProps;
  var horizontal = maxCounter > _constants.DEFAULT_AVATAR_GROUP_COUNTER ? true : horizontalProps;
  var childrenComponents = children ? [].concat(children) : [];
  childrenComponents = childrenComponents.filter(function (child) {
    return child && child.type && (child.type.displayName === 'zmp-avatar' || child.type.name === 'zmp-avatar');
  });
  var total = childrenComponents.length;
  var dataToDisplay = childrenComponents.slice(0, maxCounter).map(function (child) {
    return child.props;
  });
  var classes = (0, _utils.classNames)('avatar-group', {
    'avatar-group-horizontal': horizontal
  }, className);
  return /*#__PURE__*/_react.default.createElement("div", {
    id: id,
    ref: elRef,
    className: classes,
    style: style
  }, dataToDisplay.map(function (avatarProps, index) {
    return /*#__PURE__*/_react.default.createElement(_avatar.default, (0, _extends2.default)({
      className: "avatar-group-item",
      key: avatarProps.id || "avatar-" + index
    }, avatarProps, {
      story: false,
      online: false,
      size: 24
    }));
  }), total > maxCounter && /*#__PURE__*/_react.default.createElement(_avatar.default, {
    onClick: handleOnCounterClick,
    className: "avatar-group-item avatar-max-counter",
    size: 24
  }, "+", total - maxCounter));
});
AvatarGroup.displayName = 'zmp-avatar-group';
var _default = AvatarGroup;
exports.default = _default;
