// zmp-react/cjs/zalo-ui-kits/nav-left.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _constants = require("../../common/constants");

var _navLeft = _interopRequireDefault(require("../components/nav-left"));

var _link = _interopRequireDefault(require("../components/link"));

var _utils = require("../shared/utils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* dts-props
  id?: string | number;
  className?: string;
  style?: React.CSSProperties;
  backLink? : boolean | string;
  backLinkUrl? : string;
  backLinkForce? : boolean;
  backLinkShowText? : boolean;
  sliding? : boolean;
  COLOR_PROPS
  onBackClick? : (event?: any) => void;
  onClickBack? : (event?: any) => void;
  ref?: React.MutableRefObject<{el: HTMLElement | null}>;
  CHILDREN_PROP
*/
var ZMPNavLeft = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var id = props.id,
      style = props.style,
      children = props.children,
      backLink = props.backLink,
      backLinkUrl = props.backLinkUrl,
      backLinkForce = props.backLinkForce,
      backLinkShowText = props.backLinkShowText;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);
  (0, _react.useImperativeHandle)(ref, function () {
    return elRef.current;
  });

  var onBackClick = function onBackClick(event) {
    (0, _utils.emit)(props, 'backClick clickBack', event);
  };

  var needBackLinkText = backLinkShowText;
  var linkEl;

  if (backLink) {
    var text = backLink !== true && needBackLinkText ? backLink : undefined;
    linkEl = /*#__PURE__*/_react.default.createElement(_link.default, {
      href: backLinkUrl || '#',
      back: true,
      noLinkClass: true,
      iconZMP: _constants.ICON_CHEVRON_LEFT,
      force: backLinkForce || undefined,
      className: !text ? 'icon-only no-ripple' : undefined,
      text: text,
      onClick: onBackClick
    });
  }

  return /*#__PURE__*/_react.default.createElement(_navLeft.default, (0, _extends2.default)({
    id: id,
    style: style,
    backLink: false,
    ref: elRef
  }, extraAttrs), linkEl, children);
});
ZMPNavLeft.displayName = 'zmp-nav-left';
var _default = ZMPNavLeft;
exports.default = _default;
