// zmp-react/cjs/zalo-ui-kits/preloader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _utils = require("../shared/utils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* dts-props
  id?: string | number;
  className?: string;
  style?: React.CSSProperties;
  logo?: any
  size? : number | string;
  ref?: React.MutableRefObject<{el: HTMLElement | null}>;
  CHILDREN_PROP
*/
var Preloader = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var className = props.className,
      _props$id = props.id,
      id = _props$id === void 0 ? (0, _utils.getComponentId)() : _props$id,
      style = props.style,
      size = props.size,
      logo = props.logo;
  var extraAttrs = (0, _utils.getExtraAttrs)(props);
  var elRef = (0, _react.useRef)(null);
  (0, _react.useImperativeHandle)(ref, function () {
    return {
      el: elRef.current
    };
  });
  var preloaderStyle = {};
  var sizeComputed = size;

  if (sizeComputed && typeof sizeComputed === 'string' && sizeComputed.indexOf('px') >= 0) {
    sizeComputed = sizeComputed.replace('px', '');
  }

  if (sizeComputed) {
    preloaderStyle.width = sizeComputed + "px";
    preloaderStyle.height = sizeComputed + "px";
    preloaderStyle['--zmp-preloader-size'] = sizeComputed + "px";
  }

  if (style) (0, _utils.extend)(preloaderStyle, style || {});

  var preloaderLogo = /*#__PURE__*/_react.default.createElement("div", {
    className: "ring-spinner"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "ball-holder"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "ball"
  })));

  var innerEl = /*#__PURE__*/_react.default.createElement("svg", {
    className: "spinner",
    viewBox: "0 0 66 66",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_react.default.createElement("circle", {
    className: "spinner-path",
    fill: "transparent",
    cx: "33",
    cy: "33",
    r: "30",
    strokeWidth: "5",
    stroke: "url(#gradient-" + id + ")"
  }), /*#__PURE__*/_react.default.createElement("linearGradient", {
    id: "gradient-" + id
  }, /*#__PURE__*/_react.default.createElement("stop", {
    offset: "20%",
    stopOpacity: "1"
  }), /*#__PURE__*/_react.default.createElement("stop", {
    offset: "50%",
    stopOpacity: ".5"
  }), /*#__PURE__*/_react.default.createElement("stop", {
    offset: "100%",
    stopOpacity: "0"
  })), /*#__PURE__*/_react.default.createElement("svg", {
    className: "spinner-dot dot",
    width: "5",
    height: "5",
    viewBox: "0 0 66 66",
    xmlns: "http://www.w3.org/2000/svg",
    x: "39",
    y: "1.7"
  }, /*#__PURE__*/_react.default.createElement("circle", {
    className: "dot-path",
    cx: "33",
    cy: "33",
    r: "30"
  })));

  var classes = (0, _utils.classNames)(className, {
    'preloader-with-logo': logo
  }, 'zmp-preloader');
  return /*#__PURE__*/_react.default.createElement("span", (0, _extends2.default)({
    id: id,
    style: preloaderStyle,
    className: classes,
    ref: elRef
  }, extraAttrs), logo ? preloaderLogo : innerEl, logo && /*#__PURE__*/_react.default.createElement("div", {
    className: "preloader-logo-container"
  }, /*#__PURE__*/_react.default.createElement("img", {
    src: logo,
    className: "logo",
    alt: ""
  })));
});
Preloader.displayName = 'zmp-preloader';
var _default = Preloader;
exports.default = _default;
