// zmp-react/cjs/zalo-ui-kits/sheet.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _sheet = _interopRequireDefault(require("../components/sheet"));

var _icon = _interopRequireDefault(require("./icon"));

var _link = _interopRequireDefault(require("../components/link"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* dts-imports
import { Sheet } from 'zmp-core/types';
*/

/* dts-props
  id?: string | number;
  className?: string;
  style?: React.CSSProperties;
  opened? : boolean
  animate? : boolean
  top? : boolean
  bottom? : boolean
  position? : string
  backdrop? : boolean
  backdropEl? : string | object
  closeByBackdropClick? : boolean
  closeByOutsideClick? : boolean
  closeOnEscape? : boolean
  push? : boolean
  swipeToClose? : boolean
  swipeToStep? : boolean
  swipeHandler? : string | object
  containerEl? : string | object
  closeButton? : boolean
  title? : string
  subTitle? : string
  COLOR_PROPS
  onSheetStepProgress? : (instance?: Sheet.Sheet, progress?: any) => void
  onSheetStepOpen? : (instance?: Sheet.Sheet) => void
  onSheetStepClose? : (instance?: Sheet.Sheet) => void
  onSheetOpen? : (instance?: Sheet.Sheet) => void
  onSheetOpened? : (instance?: Sheet.Sheet) => void
  onSheetClose? : (instance?: Sheet.Sheet) => void
  onSheetClosed? : (instance?: Sheet.Sheet) => void
  ref?: React.MutableRefObject<{el: HTMLElement | null; zmpSheet: () => Sheet.Sheet;}>;
  CHILDREN_PROP
*/
var ZMPSheet = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var sheetRef = (0, _react.useRef)(null);
  var title = props.title,
      subtitle = props.subtitle,
      closeButton = props.closeButton,
      children = props.children,
      others = (0, _objectWithoutPropertiesLoose2.default)(props, ["title", "subtitle", "closeButton", "children"]);
  (0, _react.useImperativeHandle)(ref, function () {
    return sheetRef.current;
  });
  var headerEl = null;

  if (title || closeButton) {
    headerEl = /*#__PURE__*/_react.default.createElement("div", {
      slot: "fixed",
      className: "zmp-sheet-default-toolbar"
    }, closeButton && /*#__PURE__*/_react.default.createElement(_link.default, {
      sheetClose: true,
      className: "zmp-sheet-default-close-button no-ripple"
    }, /*#__PURE__*/_react.default.createElement(_icon.default, {
      zmp: "zi-close"
    })), title && /*#__PURE__*/_react.default.createElement("div", {
      className: "zmp-sheet-default-title-container"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "zmp-sheet-default-title"
    }, title), subtitle && /*#__PURE__*/_react.default.createElement("div", {
      className: "zmp-sheet-default-subtitle"
    }, subtitle)));
  }

  return /*#__PURE__*/_react.default.createElement(_sheet.default, (0, _extends2.default)({
    ref: sheetRef
  }, others), headerEl, children);
});
ZMPSheet.displayName = 'zmp-sheet';
var _default = ZMPSheet;
exports.default = _default;
